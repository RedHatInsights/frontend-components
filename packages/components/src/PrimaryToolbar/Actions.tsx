/* eslint-disable rulesdir/forbid-pf-relative-imports */
import React, { Fragment, useState } from 'react';
import {
  Button,
  Divider,
  Dropdown,
  DropdownItem,
  DropdownList,
  DropdownProps,
  MenuToggle,
  MenuToggleProps,
  ToolbarItem,
} from '@patternfly/react-core';

import EllipsisVIcon from '@patternfly/react-icons/dist/dynamic/icons/ellipsis-v-icon';

import { DownloadButton, DownloadButtonProps } from '../DownloadButton';
import classNames from 'classnames';

type ActionObject = {
  key?: string;
  label: React.ReactNode;
  value?: number | string;
  onClick?: (
    event: MouseEvent | React.MouseEvent<any, MouseEvent> | React.KeyboardEvent<Element>,
    action: ActionObject,
    key: string | number
  ) => void;
  props?: { [key: string]: any; className?: string };
};

export type ActionsType = React.ReactNode | ActionObject;

export interface ActionsProps {
  actions?: ActionsType[];
  onSelect?: DropdownProps['onSelect'];
  overflowActions?: ActionsType[];
  dropdownProps?: DropdownProps;
  kebabToggleProps?: MenuToggleProps;
  exportConfig?: DownloadButtonProps;
}

function isActionObject(node: React.ReactNode | ActionObject): node is ActionObject {
  return (node as ActionObject).label !== undefined;
}

export const overflowActionsMapper = (action: ActionsType, key: string | number) => {
  const internalAction = action as ActionObject;
  return (
    <DropdownItem
      {...internalAction.props}
      className="ins-c-primary-toolbar__overflow-actions"
      key={internalAction.value || internalAction.key || `${key}-overflow`}
      component={(internalAction.props && internalAction.props.component) || React.isValidElement(internalAction.label || action) ? 'div' : 'button'}
      onClick={(e) => internalAction.onClick && internalAction.onClick(e, internalAction, key)}
    >
      {/* FIXME: fix typings */}
      {internalAction.label || (action as React.ReactNode)}
    </DropdownItem>
  );
};

export const actionPropsGenerator = (action: ActionsType, key: string | number) => {
  const onClick =
    typeof action === 'object' && typeof action !== null && typeof (action as { onClick?: any })?.onClick === 'function'
      ? (e: MouseEvent | React.MouseEvent<any, MouseEvent> | React.KeyboardEvent<Element>) => (action as { onClick?: any })?.onClick(e, action, key)
      : undefined;

  return {
    ...(action as ActionObject)?.props,
    onClick,
    component: (action as ActionObject)?.props?.component || (React.isValidElement((action as ActionObject).label || action) ? 'div' : 'button'),
    children: (typeof action === 'object' && typeof action !== null ? (action as { label?: React.ReactNode })?.label : action) as React.ReactNode,
  };
};

const Actions: React.FunctionComponent<ActionsProps> = ({
  actions = [],
  overflowActions = [],
  onSelect = () => undefined,
  dropdownProps = {},
  exportConfig,
}) => {
  const [isOpen, setIsOpen] = useState(false);

  const toggleOpen = (isOpen: boolean) => setIsOpen(isOpen);

  const [firstAction, ...restActions] = actions;

  const dropdownItems = [
    ...(firstAction
      ? [
          <DropdownItem
            key="first-action"
            {...actionPropsGenerator(firstAction, 'first-action')}
            className={classNames('ins-c-primary-toolbar__first-action', isActionObject(firstAction) ? firstAction?.props?.className : undefined)}
          />,
        ]
      : []),
    ...restActions.map((action, key) => (
      <DropdownItem key={(action as ActionObject)?.key || (action as ActionObject)?.props?.key || key} {...actionPropsGenerator(action, key)} />
    )),
    ...(actions.length > 0 && overflowActions.length > 0
      ? [<Divider key="separator" className="ins-c-primary-toolbar__overflow-actions-separator" />]
      : []),
    ...overflowActions.map((action, key) => overflowActionsMapper(action as ActionObject, key)),
  ];

  return (
    <Fragment>
      {firstAction && (
        <ToolbarItem className="ins-c-primary-toolbar__first-action pf-m-spacer-sm">
          {isActionObject(firstAction) ? (
            <Button
              ouiaId={`${firstAction.label}`}
              {...firstAction.props}
              onClick={firstAction.onClick || (firstAction.props && firstAction.props.onClick) || undefined}
            >
              {firstAction.label}
            </Button>
          ) : (
            firstAction
          )}
        </ToolbarItem>
      )}
      {exportConfig && (exportConfig.extraItems || exportConfig.onSelect) && (
        <ToolbarItem className="pf-m-spacer-sm">
          <DownloadButton {...exportConfig} />
        </ToolbarItem>
      )}
      {(actions?.length > 0 || overflowActions.length > 0) && (
        <ToolbarItem className={`${actions.length <= 1 ? 'ins-m-actions--empty' : ''} ins-c-primary-toolbar__actions pf-m-spacer-sm`}>
          <Dropdown
            {...dropdownProps}
            isOpen={isOpen}
            onSelect={(...props) => {
              onSelect?.(...props);
              toggleOpen(false);
            }}
            onOpenChange={setIsOpen}
            ouiaId="BulkActionsList"
            toggle={(toggleRef) => (
              <MenuToggle
                ref={toggleRef}
                aria-label="kebab dropdown toggle"
                variant="plain"
                onClick={() => setIsOpen((prev) => !prev)}
                isExpanded={isOpen}
                data-ouia-component-id="BulkActionsToggle"
              >
                <EllipsisVIcon />
              </MenuToggle>
            )}
          >
            <DropdownList>{dropdownItems}</DropdownList>
          </Dropdown>
        </ToolbarItem>
      )}
    </Fragment>
  );
};

export default Actions;
