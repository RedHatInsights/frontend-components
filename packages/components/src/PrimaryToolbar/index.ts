export { default } from './PrimaryToolbar';
export { default as PrimaryToolbar } from './PrimaryToolbar';
export { default as Actions } from './Actions';
export { default as SortBy } from './SortBy';
export * from './SortBy';
export * from './PrimaryToolbar';
