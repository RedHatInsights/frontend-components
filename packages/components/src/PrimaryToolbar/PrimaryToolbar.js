import { Toolbar, ToolbarContent, ToolbarGroup, ToolbarItem, Pagination, Button, ToolbarExpandIconWrapper } from '@patternfly/react-core';
import { AngleDownIcon, AngleRightIcon } from '@patternfly/react-icons';
import React, { Component } from 'react';

import Actions from './Actions';
import { BulkSelect } from '../BulkSelect';
import { ConditionalFilter } from '../ConditionalFilter';
import { DownloadButton } from '../DownloadButton';
import { FilterChips } from '../FilterChips';
import PropTypes from 'prop-types';
import SortBy from './SortBy';
import { SortByDirection } from '@patternfly/react-table';
import './primary-toolbar.scss';

class PrimaryToolbar extends Component {
  render() {
    const {
      id,
      className,
      toggleIsExpanded,
      bulkSelect,
      filterConfig,
      dedicatedAction,
      actionsConfig,
      sortByConfig,
      pagination,
      activeFiltersConfig,
      children,
      exportConfig,
      expandAll,
      useMobileLayout,
      ...props
    } = this.props;
    const overflowActions = [
      ...(sortByConfig
        ? [
            {
              label: 'Sort order ASC',
              props: { isDisabled: sortByConfig.direction === SortByDirection.asc },
              onClick: (e) => sortByConfig.onSortChange && sortByConfig.onSortChange(e, SortByDirection.asc),
            },
            {
              label: 'Sort order DESC',
              props: { isDisabled: sortByConfig.direction === SortByDirection.desc },
              onClick: (e) => sortByConfig.onSortChange && sortByConfig.onSortChange(e, SortByDirection.desc),
            },
          ]
        : []),
    ];
    return (
      <Toolbar
        {...props}
        className={`${className || ''} ins-c-primary-toolbar`}
        toggleIsExpanded={toggleIsExpanded}
        id={id || 'ins-primary-data-toolbar'}
        ouiaId="PrimaryToolbar"
      >
        <ToolbarContent>
          {(expandAll || bulkSelect || filterConfig || dedicatedAction) && (
            <ToolbarGroup className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg" variant="filter-group">
              {expandAll && (
                <ToolbarItem>
                  {React.isValidElement(expandAll) ? (
                    expandAll
                  ) : (
                    <Button
                      variant="plain"
                      aria-label={`${expandAll.isAllExpanded ? 'Collapse' : 'Expand'} all`}
                      onClick={(e) => expandAll.onClick(e, !expandAll.isAllExpanded)}
                      ouiaId="ExpandCollapseAll"
                    >
                      <ToolbarExpandIconWrapper>{expandAll.isAllExpanded ? <AngleDownIcon /> : <AngleRightIcon />}</ToolbarExpandIconWrapper>
                    </Button>
                  )}
                </ToolbarItem>
              )}
              {bulkSelect && (
                <ToolbarItem>{React.isValidElement(bulkSelect) ? bulkSelect : <BulkSelect ouiaId="BulkSelect" {...bulkSelect} />}</ToolbarItem>
              )}
              {filterConfig && (
                <ToolbarItem className="ins-c-primary-toolbar__filter">
                  {React.isValidElement(filterConfig) ? filterConfig : <ConditionalFilter useMobileLayout={useMobileLayout} {...filterConfig} />}
                </ToolbarItem>
              )}
              {dedicatedAction && <ToolbarItem>{dedicatedAction}</ToolbarItem>}
            </ToolbarGroup>
          )}
          {React.isValidElement(actionsConfig)
            ? actionsConfig
            : ((actionsConfig && actionsConfig.actions && actionsConfig.actions.length > 0) || sortByConfig || exportConfig) && (
                <Actions {...(actionsConfig || {})} exportConfig={exportConfig} overflowActions={overflowActions} />
              )}
          {sortByConfig && (
            <ToolbarItem className="ins-c-primary-toolbar__sort-by">
              {React.isValidElement(sortByConfig) ? sortByConfig : <SortBy {...sortByConfig} />}
            </ToolbarItem>
          )}
          {children}
          {pagination && (
            <ToolbarItem className="ins-c-primary-toolbar__pagination">
              {React.isValidElement(pagination) ? pagination : <Pagination isCompact ouiaId="CompactPagination" {...pagination} />}
            </ToolbarItem>
          )}
        </ToolbarContent>
        {activeFiltersConfig && React.isValidElement(activeFiltersConfig) ? (
          <ToolbarContent>
            <ToolbarItem>{activeFiltersConfig}</ToolbarItem>
          </ToolbarContent>
        ) : (
          activeFiltersConfig !== undefined &&
          (activeFiltersConfig.filters.length > 0 || activeFiltersConfig.showDeleteButton === true) && (
            <ToolbarContent>
              <ToolbarItem>
                <FilterChips {...activeFiltersConfig} />
              </ToolbarItem>
            </ToolbarContent>
          )
        )}
      </Toolbar>
    );
  }
}

PrimaryToolbar.propTypes = {
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  className: PropTypes.string,
  toggleIsExpanded: PropTypes.func,
  /** @reference [BulkSelect props](/components/BulkSelect) */
  bulkSelect: PropTypes.shape(BulkSelect.propTypes),
  /** @reference [ConditionalFilter props](/components/ConditionalFilter) */
  filterConfig: PropTypes.shape(ConditionalFilter.propTypes),
  dedicatedAction: PropTypes.node,
  /** @reference [PF pagination props](https://www.patternfly.org/v4/components/pagination#pagination) */
  pagination: PropTypes.shape(Pagination.propTypes),
  /** @reference [SortBy props](/components/SortBy) */
  sortByConfig: PropTypes.shape(SortBy.propTypes),
  /** @reference [DownloadButton props](/components/DownloadButton) */
  exportConfig: PropTypes.shape(DownloadButton.propTypes),
  /** @reference [FilterChips props](/components/FilterChips) */
  activeFiltersConfig: PropTypes.shape(FilterChips.propTypes),
  children: PropTypes.node,
  actionsConfig: PropTypes.shape({
    actions: Actions.propTypes.actions,
    dropdownProps: Actions.propTypes.dropdownProps,
    onSelect: Actions.propTypes.onSelect,
  }),
  expandAll: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.shape({
      onClick: PropTypes.func,
      isAllExpanded: PropTypes.bool,
    }),
  ]),
  /** Use PF [toolbar toggle component for conditional filter](https://www.patternfly.org/v4/components/toolbar/react/component-managed-toggle-groups/) */
  useMobileLayout: PropTypes.bool,
};

export default PrimaryToolbar;
