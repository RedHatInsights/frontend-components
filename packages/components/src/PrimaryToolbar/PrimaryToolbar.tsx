import React from 'react';
import {
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem,
  Pagination,
  Button,
  ToolbarExpandIconWrapper,
  PaginationProps,
  ButtonProps,
} from '@patternfly/react-core';
import { AngleDownIcon, AngleRightIcon } from '@patternfly/react-icons';
import { SortByDirection } from '@patternfly/react-table';

import Actions, { ActionsProps } from './Actions';
import { BulkSelect, BulkSelectProps } from '../BulkSelect';
import { ConditionalFilter } from '../ConditionalFilter';
import { DownloadButtonProps } from '../DownloadButton';
import { FilterChips, FilterChipsProps } from '../FilterChips';
import SortBy, { SortByProps } from './SortBy';

import './primary-toolbar.scss';

export type PrimaryToolbarExpandAllObject = {
  onClick: (event: React.MouseEvent<Element, MouseEvent>, isExpanded: boolean) => void;
  isAllExpanded?: boolean;
  buttonProps?: ButtonProps;
  isDisabled?: boolean;
};

export interface PrimaryToolbarProps {
  id?: string | number;
  className?: string;
  toggleIsExpanded?: () => 'bla';
  /** @reference [BulkSelect props](/components/BulkSelect) */
  bulkSelect?: BulkSelectProps;
  /** @reference [ConditionalFilter props](/components/ConditionalFilter)
   * TODO: use ConditionalFilterProps: https://github.com/RedHatInsights/frontend-components/pull/1379
   */
  filterConfig?: any;
  dedicatedAction?: React.ReactNode;
  /** @reference [PF pagination props](https://www.patternfly.org/v4/components/pagination#pagination) */
  pagination?: Omit<PaginationProps, 'ref'>;

  /** @reference [SortBy props](/components/SortBy) */
  sortByConfig: SortByProps;
  /** @reference [DownloadButton props](/components/DownloadButton) */
  exportConfig: DownloadButtonProps;

  /** @reference [FilterChips props](/components/FilterChips)
   */
  activeFiltersConfig?: FilterChipsProps;
  actionsConfig?: ActionsProps;
  expandAll?: React.ReactNode | PrimaryToolbarExpandAllObject;

  /** Use PF [toolbar toggle component for conditional filter](https://www.patternfly.org/v4/components/toolbar/react/component-managed-toggle-groups/) */
  useMobileLayout: boolean;
}

function isPrimaryToolbarExpandAllObject(node: React.ReactNode | PrimaryToolbarExpandAllObject): node is PrimaryToolbarExpandAllObject {
  return !React.isValidElement(node);
}

const PrimaryToolbar: React.FunctionComponent<PrimaryToolbarProps> = ({
  id,
  className,
  toggleIsExpanded,
  bulkSelect,
  filterConfig,
  dedicatedAction,
  actionsConfig,
  sortByConfig,
  pagination,
  activeFiltersConfig,
  children,
  exportConfig,
  expandAll,
  useMobileLayout,
  ...props
}) => {
  const overflowActions = [
    ...(sortByConfig
      ? [
          {
            label: 'Sort order ASC',
            props: { isDisabled: sortByConfig.direction === SortByDirection.asc },
            onClick: (e: React.MouseEvent) => sortByConfig.onSortChange && sortByConfig.onSortChange(e, SortByDirection.asc),
          },
          {
            label: 'Sort order DESC',
            props: { isDisabled: sortByConfig.direction === SortByDirection.desc },
            onClick: (e: React.MouseEvent) => sortByConfig.onSortChange && sortByConfig.onSortChange(e, SortByDirection.desc),
          },
        ]
      : []),
  ];
  return (
    <Toolbar
      {...props}
      className={`${className || ''} ins-c-primary-toolbar`}
      toggleIsExpanded={toggleIsExpanded}
      id={id ? `${id}` : 'ins-primary-data-toolbar'}
      ouiaId="PrimaryToolbar"
    >
      <ToolbarContent>
        {(expandAll || bulkSelect || filterConfig || dedicatedAction) && (
          <ToolbarGroup className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg" variant="filter-group">
            {expandAll && (
              <ToolbarItem>
                {isPrimaryToolbarExpandAllObject(expandAll) ? (
                  <Button
                    {...expandAll.buttonProps}
                    variant="plain"
                    aria-label={`${expandAll.isAllExpanded ? 'Collapse' : 'Expand'} all`}
                    onClick={(e) => expandAll.onClick(e, !expandAll.isAllExpanded)}
                    ouiaId="ExpandCollapseAll"
                    isDisabled={expandAll.isDisabled}
                  >
                    <ToolbarExpandIconWrapper>{expandAll.isAllExpanded ? <AngleDownIcon /> : <AngleRightIcon />}</ToolbarExpandIconWrapper>
                  </Button>
                ) : (
                  expandAll
                )}
              </ToolbarItem>
            )}
            {bulkSelect && (
              <ToolbarItem>{React.isValidElement(bulkSelect) ? bulkSelect : <BulkSelect ouiaId="BulkSelect" {...bulkSelect} />}</ToolbarItem>
            )}
            {filterConfig && (
              <ToolbarItem className="ins-c-primary-toolbar__filter">
                {React.isValidElement(filterConfig) ? filterConfig : <ConditionalFilter useMobileLayout={useMobileLayout} {...filterConfig} />}
              </ToolbarItem>
            )}
            {dedicatedAction && <ToolbarItem>{dedicatedAction}</ToolbarItem>}
          </ToolbarGroup>
        )}
        {React.isValidElement(actionsConfig)
          ? actionsConfig
          : ((actionsConfig && actionsConfig.actions && actionsConfig.actions.length > 0) || sortByConfig || exportConfig) && (
              <Actions {...(actionsConfig || {})} exportConfig={exportConfig} overflowActions={overflowActions} />
            )}
        {sortByConfig && (
          <ToolbarItem className="ins-c-primary-toolbar__sort-by">
            {React.isValidElement(sortByConfig) ? sortByConfig : <SortBy {...sortByConfig} />}
          </ToolbarItem>
        )}
        {children}
        {pagination && (
          <ToolbarItem className="ins-c-primary-toolbar__pagination">
            {React.isValidElement(pagination) ? pagination : <Pagination isCompact ouiaId="CompactPagination" {...pagination} />}
          </ToolbarItem>
        )}
      </ToolbarContent>
      {activeFiltersConfig && React.isValidElement(activeFiltersConfig) ? (
        <ToolbarContent>
          <ToolbarItem>{activeFiltersConfig}</ToolbarItem>
        </ToolbarContent>
      ) : (
        typeof activeFiltersConfig?.filters !== 'undefined' &&
        (activeFiltersConfig.filters.length > 0 || activeFiltersConfig.showDeleteButton === true) && (
          <ToolbarContent>
            <ToolbarItem>
              <FilterChips {...activeFiltersConfig} />
            </ToolbarItem>
          </ToolbarContent>
        )
      )}
    </Toolbar>
  );
};

export default PrimaryToolbar;
