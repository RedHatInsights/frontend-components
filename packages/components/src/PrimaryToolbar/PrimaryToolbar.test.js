import React from 'react';
import PrimaryToolbar from './PrimaryToolbar';
import { Button } from '@patternfly/react-core/dist/dynamic/components/Button';
import { act, render } from '@testing-library/react';

const onActionClicked = jest.fn();
const onActionSelect = jest.fn();
const config = {
  useMobileLayout: true,
  pagination: {
    itemCount: 100,
    page: 1,
    perPage: 10,
    onSetPage: jest.fn(),
    onPerPageSelect: jest.fn(),
  },
  exportConfig: {
    onSelect: jest.fn(),
  },
  actionsConfig: {
    actions: [
      <Button key="first" onClick={onActionClicked}>
        {' '}
        Some action
      </Button>,
      <Button key="second" onClick={onActionClicked}>
        Another button
      </Button>,
      {
        label: 'Or objects',
        onClick: onActionClicked,
      },
      'or plain string',
    ],
    onSelect: onActionSelect,
    dropdownProps: { className: 'custom-class' },
  },
  sortByConfig: {
    direction: 'asc',
    onSortChange: jest.fn(),
  },
  bulkSelect: {
    items: [
      {
        title: 'Some action',
        onClick: jest.fn(),
      },
    ],
    checked: false,
    onSelect: jest.fn(),
  },
  filterConfig: {
    items: [
      {
        type: 'text',
        label: 'First filter',
      },
      {
        label: 'Second filter',
        type: 'checkbox',
        filterValues: {
          items: [{ label: 'Some checkbox' }],
        },
      },
    ],
  },
  activeFiltersConfig: {
    filters: [
      {
        category: 'Some',
        chips: [
          {
            name: 'something',
          },
          {
            name: 'something 2',
          },
        ],
      },
      {
        category: 'Another',
        chips: [
          {
            name: 'One chip',
          },
        ],
      },
      {
        name: 'Something else',
      },
    ],
    onDelete: jest.fn(),
  },
};

describe('PrimaryToolbar', () => {
  describe('should render', () => {
    it('no data', () => {
      const { container } = render(<PrimaryToolbar />);
      expect(container).toMatchSnapshot();
    });

    it('elements instead of data', () => {
      const elementsConfig = Object.keys(config)
        .map((key) => ({
          [key]: <div>Loading...</div>,
        }))
        .reduce((acc, curr) => ({ ...acc, ...curr }), {});

      const { container } = render(<PrimaryToolbar {...elementsConfig} />);
      expect(container).toMatchSnapshot();
    });

    describe('with data', () => {
      it('full config', () => {
        const { container } = render(<PrimaryToolbar {...config} />);
        expect(container).toMatchSnapshot();
      });

      it('only - bulk select', () => {
        const { container } = render(<PrimaryToolbar bulkSelect={config.bulkSelect} />);
        expect(container).toMatchSnapshot();
      });

      it('only - filterConfig', () => {
        const { container } = render(<PrimaryToolbar filterConfig={config.filterConfig} />);
        expect(container).toMatchSnapshot();
      });
    });

    it('custom className', () => {
      const { container } = render(<PrimaryToolbar className="custom-classname" />);
      expect(container).toMatchSnapshot();
    });

    it('custom id', () => {
      const { container } = render(<PrimaryToolbar id="custom-id" />);
      expect(container).toMatchSnapshot();
    });

    it('wrong actionsConfig', () => {
      // eslint-disable-next-line no-unused-vars
      const { actionsConfig, ...rest } = config;
      const { container } = render(<PrimaryToolbar {...rest} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('API', () => {
    it('should call DESC sort', () => {
      const { container } = render(<PrimaryToolbar {...config} sortByConfig={{ ...config.sortByConfig, direction: 'desc' }} />);
      act(() => {
        container.querySelectorAll('.ins-c-primary-toolbar__actions button.pf-v6-c-menu-toggle')[0].click();
      });
      act(() => {
        container.querySelector('.ins-c-primary-toolbar__overflow-actions.pf-v6-c-menu__item').click();
      });
      expect(config.sortByConfig.onSortChange).toHaveBeenCalled();
      // FIXME: assertions are not working
      // expect(config.sortByConfig.onSortChange).toHaveBeenCalledWith('asc');
    });

    it('should call ASC sort', () => {
      const { container } = render(<PrimaryToolbar {...config} />);
      act(() => {
        container.querySelector('.ins-c-primary-toolbar__actions button.pf-v6-c-menu-toggle').click();
      });
      act(() => {
        container.querySelector('.ins-c-primary-toolbar__overflow-actions.pf-v6-c-menu__item').click();
      });
      expect(config.sortByConfig.onSortChange).toHaveBeenCalled();
      // FIXME: assertions are not working
      // expect(config.sortByConfig.onSortChange).toHaveBeenCalledWith('desc');
    });
  });
});
