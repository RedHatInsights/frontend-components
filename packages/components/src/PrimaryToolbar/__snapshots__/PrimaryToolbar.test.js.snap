// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryToolbar should render custom className 1`] = `
<Toolbar
  className="custom-classname ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  />
</Toolbar>
`;

exports[`PrimaryToolbar should render custom id 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="custom-id"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  />
</Toolbar>
`;

exports[`PrimaryToolbar should render elements instead of data 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
  useMobileLayout={
    <div>
      Loading...
    </div>
  }
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
      variant="filter-group"
    >
      <ToolbarItem>
        <div>
          Loading...
        </div>
      </ToolbarItem>
      <ToolbarItem
        className="ins-c-primary-toolbar__filter"
      >
        <div>
          Loading...
        </div>
      </ToolbarItem>
    </ForwardRef>
    <div>
      Loading...
    </div>
    <ToolbarItem
      className="ins-c-primary-toolbar__sort-by"
    >
      <div>
        Loading...
      </div>
    </ToolbarItem>
    <ToolbarItem
      className="ins-c-primary-toolbar__pagination"
    >
      <div>
        Loading...
      </div>
    </ToolbarItem>
  </ToolbarContent>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ToolbarItem>
      <div>
        Loading...
      </div>
    </ToolbarItem>
  </ToolbarContent>
</Toolbar>
`;

exports[`PrimaryToolbar should render no data 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  />
</Toolbar>
`;

exports[`PrimaryToolbar should render with data full config 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
  useMobileLayout={true}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
      variant="filter-group"
    >
      <ToolbarItem>
        <BulkSelect
          checked={false}
          items={
            [
              {
                "onClick": [MockFunction],
                "title": "Some action",
              },
            ]
          }
          onSelect={[MockFunction]}
        />
      </ToolbarItem>
      <ToolbarItem
        className="ins-c-primary-toolbar__filter"
      >
        <ConditionalFilter
          items={
            [
              {
                "label": "First filter",
                "type": "text",
              },
              {
                "filterValues": {
                  "items": [
                    {
                      "label": "Some checkbox",
                    },
                  ],
                },
                "label": "Second filter",
                "type": "checkbox",
              },
            ]
          }
        />
      </ToolbarItem>
    </ForwardRef>
    <Actions
      actions={
        [
          <Button
            onClick={[MockFunction]}
          >
             
            Some action
          </Button>,
          <Button
            onClick={[MockFunction]}
          >
            Another button
          </Button>,
          {
            "label": "Or objects",
            "onClick": [MockFunction],
          },
          "or plain string",
        ]
      }
      dropdownProps={
        {
          "className": "custom-class",
        }
      }
      exportConfig={
        {
          "onSelect": [MockFunction],
        }
      }
      onSelect={[MockFunction]}
      overflowActions={
        [
          {
            "label": "Sort order ASC",
            "onClick": [Function],
            "props": {
              "isDisabled": true,
            },
          },
          {
            "label": "Sort order DESC",
            "onClick": [Function],
            "props": {
              "isDisabled": false,
            },
          },
        ]
      }
    />
    <ToolbarItem
      className="ins-c-primary-toolbar__sort-by"
    >
      <SortBy
        direction="asc"
        onSortChange={[MockFunction]}
      />
    </ToolbarItem>
    <ToolbarItem
      className="ins-c-primary-toolbar__pagination"
    >
      <Pagination
        isCompact={true}
        itemCount={100}
        onPerPageSelect={[MockFunction]}
        onSetPage={[MockFunction]}
        ouiaId="CompactPagination"
        page={1}
        perPage={10}
      />
    </ToolbarItem>
  </ToolbarContent>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ToolbarItem>
      <FilterChips
        filters={
          [
            {
              "category": "Some",
              "chips": [
                {
                  "name": "something",
                },
                {
                  "name": "something 2",
                },
              ],
            },
            {
              "category": "Another",
              "chips": [
                {
                  "name": "One chip",
                },
              ],
            },
            {
              "name": "Something else",
            },
          ]
        }
        onDelete={[MockFunction]}
      />
    </ToolbarItem>
  </ToolbarContent>
</Toolbar>
`;

exports[`PrimaryToolbar should render with data only - bulk select 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
      variant="filter-group"
    >
      <ToolbarItem>
        <BulkSelect
          checked={false}
          items={
            [
              {
                "onClick": [MockFunction],
                "title": "Some action",
              },
            ]
          }
          onSelect={[MockFunction]}
        />
      </ToolbarItem>
    </ForwardRef>
  </ToolbarContent>
</Toolbar>
`;

exports[`PrimaryToolbar should render with data only - filterConfig 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
      variant="filter-group"
    >
      <ToolbarItem
        className="ins-c-primary-toolbar__filter"
      >
        <ConditionalFilter
          items={
            [
              {
                "label": "First filter",
                "type": "text",
              },
              {
                "filterValues": {
                  "items": [
                    {
                      "label": "Some checkbox",
                    },
                  ],
                },
                "label": "Second filter",
                "type": "checkbox",
              },
            ]
          }
        />
      </ToolbarItem>
    </ForwardRef>
  </ToolbarContent>
</Toolbar>
`;

exports[`PrimaryToolbar should render wrong actionsConfig 1`] = `
<Toolbar
  className=" ins-c-primary-toolbar"
  id="ins-primary-data-toolbar"
  ouiaId="PrimaryToolbar"
  toggleIsExpanded={[Function]}
  useMobileLayout={true}
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      className="ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
      variant="filter-group"
    >
      <ToolbarItem>
        <BulkSelect
          checked={false}
          items={
            [
              {
                "onClick": [MockFunction],
                "title": "Some action",
              },
            ]
          }
          onSelect={[MockFunction]}
        />
      </ToolbarItem>
      <ToolbarItem
        className="ins-c-primary-toolbar__filter"
      >
        <ConditionalFilter
          items={
            [
              {
                "label": "First filter",
                "type": "text",
              },
              {
                "filterValues": {
                  "items": [
                    {
                      "label": "Some checkbox",
                    },
                  ],
                },
                "label": "Second filter",
                "type": "checkbox",
              },
            ]
          }
        />
      </ToolbarItem>
    </ForwardRef>
    <Actions
      exportConfig={
        {
          "onSelect": [MockFunction],
        }
      }
      overflowActions={
        [
          {
            "label": "Sort order ASC",
            "onClick": [Function],
            "props": {
              "isDisabled": true,
            },
          },
          {
            "label": "Sort order DESC",
            "onClick": [Function],
            "props": {
              "isDisabled": false,
            },
          },
        ]
      }
    />
    <ToolbarItem
      className="ins-c-primary-toolbar__sort-by"
    >
      <SortBy
        direction="asc"
        onSortChange={[MockFunction]}
      />
    </ToolbarItem>
    <ToolbarItem
      className="ins-c-primary-toolbar__pagination"
    >
      <Pagination
        isCompact={true}
        itemCount={100}
        onPerPageSelect={[MockFunction]}
        onSetPage={[MockFunction]}
        ouiaId="CompactPagination"
        page={1}
        perPage={10}
      />
    </ToolbarItem>
  </ToolbarContent>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ToolbarItem>
      <FilterChips
        filters={
          [
            {
              "category": "Some",
              "chips": [
                {
                  "name": "something",
                },
                {
                  "name": "something 2",
                },
              ],
            },
            {
              "category": "Another",
              "chips": [
                {
                  "name": "One chip",
                },
              ],
            },
            {
              "name": "Something else",
            },
          ]
        }
        onDelete={[MockFunction]}
      />
    </ToolbarItem>
  </ToolbarContent>
</Toolbar>
`;
